zerver/views/streams.py

```diff
@@ -591,32 +593,29 @@ def add_subscriptions_backend(
         principals = []
 
     setting_groups_dict = {}
-    if can_remove_subscribers_group is not None:
-        setting_value = parse_group_setting_value(
-            can_remove_subscribers_group, "can_remove_subscribers_group"
-        )
-        permission_configuration = Stream.stream_permission_group_settings[
-            "can_remove_subscribers_group"
-        ]
-        can_remove_subscribers_group_value = access_user_group_for_setting(
-            setting_value,
-            user_profile,
-            setting_name="can_remove_subscribers_group",
-            permission_configuration=permission_configuration,
-        )
-        setting_groups_dict[can_remove_subscribers_group_value.id] = setting_value
-    else:
-        can_remove_subscribers_group_default_name = Stream.stream_permission_group_settings[
-            "can_remove_subscribers_group"
-        ].default_group_name
-        can_remove_subscribers_group_value = NamedUserGroup.objects.get(
-            name=can_remove_subscribers_group_default_name,
-            realm=user_profile.realm,
-            is_system_group=True,
-        )
-        setting_groups_dict[can_remove_subscribers_group_value.id] = (
-            can_remove_subscribers_group_value.id
-        )
+    group_settings_map = {}
+    request_settings_dict = locals()
+    for setting_name, permission_configuration in Stream.stream_permission_group_settings.items():
+        assert setting_name in request_settings_dict
+        if request_settings_dict[setting_name] is not None:
+            setting_request_value = request_settings_dict[setting_name]
+            setting_value = parse_group_setting_value(setting_request_value, setting_name)
+            group_settings_map[setting_name] = access_user_group_for_setting(
+                setting_value,
+                user_profile,
+                setting_name=setting_name,
+                permission_configuration=permission_configuration,
+            )
+            setting_groups_dict[group_settings_map[setting_name].id] = setting_value
+        else:
+            group_settings_map[setting_name] = NamedUserGroup.objects.get(
+                name=permission_configuration.default_group_name,
+                realm=user_profile.realm,
+                is_system_group=True,
+            )
+            setting_groups_dict[group_settings_map[setting_name].id] = group_settings_map[
+                setting_name
+            ].id
```

Comment:
This could probably be turned into a preparatory refactor commit, just changing the existing code to be generic, that we could merge even if the rest of the feature if not ready.